mobilecoind-dev-faucet/src/slam/mod.rs

```diff
@@ -0,0 +1,500 @@
+                }
+            }
+        }
+
+        // Close this so that workers that get to the end of the queue won't block
+        prepared_utxos_sender.close();
+
+        // Get the network state, which contains block height, now so that it is
+        // fresh for the slam workers
+        let network_state = {
+            let resp = mobilecoind_api_client
+                .get_network_status(&Default::default())
+                .map_err(|err| format!("Failed getting network status: {}", err))?;
+            self.block_height
+                .store(resp.network_highest_block_index, Ordering::SeqCst);
+            resp
+        };
+
+        // Now, we spawn worker threads, which build txs and
+        // submits them, with retries, in parallel
+        self.phase.store(3, Ordering::SeqCst);
+        log::info!(logger, "Slam status: {}", self.get_status().unwrap());
+        let begin_submit_time = Instant::now();
+        let prepare_time = begin_submit_time.duration_since(begin_prepare_time);
+
+        // Note: Workers are std::thread rather than tokio::task because mc_connection
+        // currently only has blocking APIs If we get an async-api for
+        // attestation and tx submission, we could change this to be tasks and just use
+        // the tokio thread pool. maybe some advantages?
+        let workers: Vec<_> = (0..params.num_threads)
+            .map(|worker_num| {
+                let this = self.clone();
+                let params = params.clone();
+                let prepared_utxos_receiver = prepared_utxos_receiver.clone();
+                let recipient = recipient.clone();
+                let account_key = account_key.clone();
+                let network_state = network_state.clone();
+                let tx_submitter = tx_submitter.clone();
+                let logger = logger.new(o! { "thread" => format!("slam-worker-{}", worker_num) });
+                std::thread::spawn(move || {
+                    this.slam_worker_entry_point(
+                        worker_num,
+                        params,
+                        prepared_utxos_receiver,
+                        recipient,
+                        account_key,
+                        network_state,
+                        tx_submitter,
+                        logger,
+                    )
+                })
+            })
+            .collect();
+
+        // Wait for each worker to finish
+        log::info!(logger, "Waiting for workers to finish");
+        for worker in workers {
+            // Note: this is blocking, we could use async_thread instead I guess. But it
+            // only blocks one thread at most since only one slam can happen at
+            // a time, so it likely doesn't matter
+            match worker.join() {
+                Ok(()) => {}
+                Err(_) => {
+                    log::error!(logger, "Slam worker error on join")
```

Comment:
Might as well log the error
```suggestion
                Err(err) => {
                    log::error!(logger, "Slam worker error on join: {}", err)
```

Reply:
I think it's not actually an error, I think it's actually a captured panic object, so it's like a `Box<dyn Any>` or something.

https://doc.rust-lang.org/stable/std/thread/type.Result.html

I am not sure that there is a good way to log that.

They seem to suggest that I can propagate the panic to this thread, but I'm not sure if that's desirable. The slam worker is pretty localized and if it panics somehow, it seems likely that the rest of the webserver can continue to work just fine. I feel like as long as the details of the panic make it into the logs (which I think they should because of the panic handler) then this error has been adequately logged.

But, curious if you think there's an advantage to propagating the panic. This server is relatively stateless and should be quick to restart. I feel like it should work fine either way.

Reply:
Ah, I see. Swallowing the panic here SGTM; let's document it.
```suggestion
                Err(_) => {
                    // This error happens if a worker thread panics. The rest of
                    // the server can still work OK, so ignore the panic.
                    // The error type is Any, so we cannot easily log it, but it
                    // was likely already logged.
                    log::error!(logger, "Slam worker error on join; see previous logs for details")
```